src/js/trove/table.js

```diff
@@ -248,6 +364,114 @@
           }, "multi-order");
         }),
 
+        'add-column': runtime.makeMethod2(function(_, colname, eltList) {
+          ffi.checkArity(3, arguments, "add-column", true);
+          runtime.checkString(colname);
+          runtime.checkList(eltList);
+          var asArray = runtime.ffi.toArray(eltList);
+          if(rows.length !== asArray.length) {
+            throw runtime.ffi.throwMessageException("column-length-mismatch");
+          }
+
+          var newRows = [];
+          for(var i = 0; i < rows.length; i += 1) {
+            newRows.push(rows[i].concat([asArray[i]]));
+          }
+          
+          return makeTable(headers.concat([colname]), newRows);
+        }),
+
+        'add-row': runtime.makeMethod1(function(_, row) {
+          ffi.checkArity(2, arguments, 'add-row', true);
+          runtime.checkRow(row);
+          var theseKeys = Object.keys(headerIndex);
+          var rowKeys = Object.keys(row.$underlyingTable.headerIndex);
+          if(theseKeys.length !== rowKeys.length) {
+            throw runtime.ffi.throwMessageException("add-row-length");
+          }
+          for(var i = 0; i < theseKeys.length; i += 1) {
+            if(headerIndex["column:" + headers[i]] !== 
+                row.$underlyingTable.headerIndex["column:" + headers[i]]) {
+              throw runtime.ffi.throwMessageException("row-name-mismatch: " + headers[i]);
+            }
+          }
+          // NOTE(joe): Here we do not copy all the sub-arrays with the
+          // existing data, we just copy the outer array containing them.
+          // This relies on the assumption that we never mutate the
+          // underlying arrays in tables, and avoids significant copying.
+          // Note also that add-column doesn't get to do this sharing because
+          // of the row-major organization of tables.
+          var newRows = rows.concat([row.$rowData]);
+          return makeTable(headers, newRows);
+        }),
+
+        'row-n': runtime.makeMethod1(function(_, row) {
+          ffi.checkArity(2, arguments, "row-n", true);
+          runtime.checkNumNonNegative(row);
+          runtime.checkNumInteger(row);
+          var rowFix = runtime.num_to_fixnum(row);
+          if(rowFix >= rows.length) {
+            throw runtime.ffi.throwMessageException("row-n-too-large");
+          }
+          return makeRow({ headerIndex: headerIndex }, rows[rowFix]);
+        }),
+
+        'all-rows': runtime.makeMethod0(function(_) {
```

Comment:
Interesting that this method returns a list of rows, whereas all-columns returns a list of lists.  It makes sense from one angle, and is inconsistent from another.  Worth calling out in the documentation.